Namespacing part 2

timwoj opened the pull request:
This should catch the remainder of the core types in the namespacing work. I wanted to get this in before we cut 3.2. There are related PRs on the bifcl and binpac repos that I will link into this one. I attached what the namespacing looks like at this point.

Next up is work on the sub-namespaces (analyzer, file_analysis, etc.)

[parsed.txt](https://github.com/zeek/zeek/files/4827972/parsed.txt)

timwoj commented:
Rebased the branch on master.

jsiwek commented on src/Val.cc:
```diff
@@ -379,61 +381,61 @@ void Val::ValDescribeReST(ODesc* d) const
 #endif
 
-bool Val::WouldOverflow(const zeek::Type* from_type, const zeek::Type* to_type, const Val* val)
+bool Val::WouldOverflow(const ::zeek::Type* from_type, const ::zeek::Type* to_type, const Val* val)
```
```suggestion
bool Val::WouldOverflow(const zeek::Type* from_type, const zeek::Type* to_type, const Val* val)
```

timwoj commented on src/Conn.h:
```diff
@@ -29,8 +29,14 @@ class RuleHdrTest;
-class RecordVal;
+
+ZEEK_FORWARD_DECLARE_NAMESPACED(Val, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(RecorVal, zeek);
```
I guess that line isn't needed at all since `RecordVal` is needed a couple of lines later in that file.

timwoj commented:
> Func and FuncPtr may be better in zeek than detail ? Would guess we want to try to keep the function call/invoke API stable for people.

But leave BuiltinFunc and ScriptFunc in detail?

> BroObj -> zeek::Obj

This brings up a larger question if we should just take this opportunity to rename all of the `Bro` names.

timwoj commented on src/List.h:
```diff
@@ -26,8 +26,9 @@
 
-// TODO: this can be removed in v3.1 when List::sort() is removed
-typedef int (*list_cmp_func)(const void* v1, const void* v2);
+//enum class [[deprecated("Remove in v4.1. Use zeek::ListOrder instead.")]] ListOrder : int { ORDERED, UNORDERED };
```
Yep, forgot to remove it when I added the deprecated version.

jsiwek commented:
> But leave BuiltinFunc and ScriptFunc in detail?

Think so.  Most people probably don't care to know/use those, they'll just want to lookup a base `Func` by name and `Invoke` it.

> This brings up a larger question if we should just take this opportunity to rename all of the `Bro` names.

Good catch, guess some of the others had already slipped by, but yeah, throwing out ideas for renaming things I see on the list:

* `Bro::UID` -> `zeek::UID`
* `BroFile` -> `zeek::File` (maybe detail, not sure)
* `BroSubstring` -> `zeek::detail::Substring`
* `BroString` -> `zeek::String`
* `Brofiler` -> `zeek::detail::ScriptProfiler`
* `binpac::Bro`* -> didn't look at where those actually live, but just Zeek* versions may make sense
* `bro_broker::`* -> probably most go in `zeek::broker::detail::`* except `zeek::broker::Manager` / `zeek::broker::mgr` (from `broker_mgr`)
* `BroExprTag` -> `zeek::detail::ExprTag`

Definitely see some other `bro_` stuff not in the list, like `bro_random()`, `bro_path()`, `bro_int_t` that warrant rename eventually.  Not that scoping all ese into this PR matters, but for ones it already touched, should just do the Bro -> Zeek conversion now.

timwoj commented:
Fixed all of the comments and did a bit of additional renaming:
zeek::detail::Func/FuncPtr -> zeek::Func/FuncPtr
zeek::detail::BroFunc -> zeek::detail::ScriptFunc (and fixed the enum)
BroObj -> Obj
BroString -> String

I left BroString.h in place because moving to String.h causes a ton of errors from compilers that are case insensitive when it comes to looking up include files when it conflicts with string.h.
